lib/learn_open/environments/base_environment.rb

```diff
@@ -0,0 +1,43 @@
+module LearnOpen
+  module Environments
+    class BaseEnvironment
+      attr_reader :io, :environment_vars, :system_adapter, :options, :logger
+      def initialize(options)
+        @io               = options.fetch(:io, LearnOpen.default_io)
+        @environment_vars = options.fetch(:environment_vars, LearnOpen.environment_vars)
+        @system_adapter   = options.fetch(:system_adapter, LearnOpen.system_adapter)
+        @logger           = options.fetch(:logger, LearnOpen.logger)
+        @options          = options
+      end
+      def open_jupyter_lab(lesson, location, editor); end
+
+      def open_lab(lesson, location, editor)
+        case lesson
+        when LearnOpen::Lessons::IosLesson
+          io.puts "You need to be on a Mac to work on iOS lessons."
+        else
+          LessonDownloader.call(lesson, location, options)
```

Comment:
This is the squint test talking, but should `LessonDownloader.call` and `DependencyInstallers.run_installers` also be wrapped in instance methods?